backend/sftp/file.go

```diff
@@ -0,0 +1,341 @@
+		if err != nil {
+			return err
+		}
+		f.sftpfile = file
+		return f.Close()
+	}
+
+	client, err := f.fileSystem.Client(f.Authority)
+	if err != nil {
+		return err
+	}
+	now := time.Now()
+	return client.Chtimes(f.Path(), now, now)
+}
+
+// Size returns the size of the remote file.
+func (f *File) Size() (uint64, error) {
+
+	client, err := f.fileSystem.Client(f.Authority)
+	if err != nil {
+		return 0, err
+	}
+	userinfo, err := client.Stat(f.Path())
+	if err != nil {
+		return 0, err
+	}
+	return uint64(userinfo.Size()), nil
+}
+
+// Location returns a vfs.Location at the location of the file. IE: if file is at
+// sftp://someuser@host.com/here/is/the/file.txt the location points to sftp://someuser@host.com/here/is/the/
+func (f *File) Location() vfs.Location {
+	return &Location{
+		fileSystem: f.fileSystem,
+		path:       path.Dir(f.path),
+		Authority:  f.Authority,
+	}
+}
+
+// Move/Copy Operations
+
+// MoveToFile puts the contents of File into the targetFile passed using File.CopyToFile.
+// If the copy succeeds, the source file is deleted. Any errors from the copy or delete are
+// returned.
+// If the given location is also sftp AND for the same user and host, the sftp Rename method is used, otherwise
+// we'll do a an io.Copy to the destination file then delete source file.
+func (f *File) MoveToFile(t vfs.File) error {
+	// sftp rename if vfs is sftp and for the same user/host
+	if f.fileSystem.Scheme() == t.Location().FileSystem().Scheme() &&
+		f.Authority.User == t.(*File).Authority.User &&
+		f.Authority.Host == t.(*File).Authority.Host {
+		//ensure destination exists before moving
+		exists, err := t.Location().Exists()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			//it doesn't matter which client we use since they are effectively the same
+			client, err := f.fileSystem.Client(f.Authority)
+			if err != nil {
+				return err
+			}
+			err = client.MkdirAll(t.Location().Path())
+			if err != nil {
+				return err
+			}
+		}
+		return f.sftpRename(t.(*File))
+	}
+
+	//otherwise do copy-delete
+	if err := f.CopyToFile(t); err != nil {
+		return err
+	}
+	return f.Delete()
+}
+
+// MoveToLocation works by creating a new file on the target location then calling MoveToFile() on it.
+func (f *File) MoveToLocation(location vfs.Location) (vfs.File, error) {
+
+	newFile, err := location.FileSystem().NewFile(location.Volume(), path.Join(location.Path(), f.Name()))
+	if err != nil {
+		return nil, err
+	}
+
+	err = f.MoveToFile(newFile)
+	if err != nil {
+		return nil, err
+	}
+	return newFile, nil
+}
+
+// CopyToFile puts the contents of File into the targetFile passed.
+func (f *File) CopyToFile(file vfs.File) error {
+
+	if err := utils.TouchCopy(file, f); err != nil {
+		return err
+	}
+	//Close target to flush and ensure that cursor isn't at the end of the file when the caller reopens for read
+	if cerr := file.Close(); cerr != nil {
+		return cerr
+	}
+	//Close file (f) reader
+	return f.Close()
```

Comment:
Shouldn't this be up to the implementer using the API? Closing internally may do something unintended, such as writing file contents remotely before file operations are complete after the copy.

```
myFile, _ := fs.NewFile(...)
otherFile, _ := fs.NewFile(...)
myFile.CopyToFile(otherFile)
myFile.Write([]byte("moar data!"))
myFile.Close()
```
This would result in the follow-up write to overwrite existing data rather than appending, and I'd have no idea unless I dug into the API source.

Reply:
In this case the CopyToFile, "opening" the file wasn't initiated by the buyer. We close internally to make this an atomic function to the user.  In your scenario, CopyToFile will jhave opened and closed the file and Write will open again, necessitating an explicit close.  Basically, we required a close on Read, Write, and Seek.

Reply:
Grrr... i wrote more expansive reply on this but lost it.  Basically, I said I'd add more docs to clarify to for these functions, explaining that we close both source and target files.  This is generally consistent with unix command line usage of `cp` or `mv` which was the behavior model.  Also noted that you can continue to append with an additional seek-to-end call 
```go
myFile, _ := fs.NewFile(...)
otherFile, _ := fs.NewFile(...)
myFile.CopyToFile(otherFile)
myFile.Seek(0,2)
myFile.Write([]byte("moar data!"))
myFile.Close()
```
See the latest commit.